x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/rules_table/rules_table_context.tsx

```diff
@@ -0,0 +1,292 @@
+  const [loadingRules, setLoadingRules] = useState<LoadingRules>({ ids: [], action: null });
+  const [page, setPage] = useState(1);
+  const [perPage, setPerPage] = useState(20);
+  const [selectedRuleIds, setSelectedRuleIds] = useState<string[]>([]);
+
+  const isActionInProgress = useMemo(() => {
+    if (loadingRules.ids.length > 0) {
+      return !(loadingRules.action === 'disable' || loadingRules.action === 'enable');
+    }
+    return false;
+  }, [loadingRules.action, loadingRules.ids.length]);
+
+  const pagination = useMemo(() => ({ page, perPage }), [page, perPage]);
+
+  // Fetch rules
+  const {
+    data: { rules, total } = { rules: [], total: 0 },
+    refetch,
+    dataUpdatedAt,
+    isFetched,
+    isFetching,
+    isLoading,
+    isRefetching,
+  } = useAllRules({
+    enabled: hasTotalRules,
+    isInMemorySorting,
+    filterOptions,
+    sortingOptions,
+    pagination,
+    refetchInterval: isRefreshOn && !isActionInProgress && autoRefreshSettings.value,
+  });
+
+  useEffect(() => {
+    // Synchronize re-fetching of rules and pre-packaged rule statuses
+    if (isFetched && isRefetching) {
+      refetchPrePackagedRulesStatus();
+    }
+  }, [isFetched, isRefetching, refetchPrePackagedRulesStatus]);
+
+  // Filter rules
+  const filteredRules = isInMemorySorting ? rules.filter(getRulesPredicate(filterOptions)) : rules;
+
+  // Paginate and sort rules
+  const rulesToDisplay = isInMemorySorting
+    ? filteredRules
+        .sort(getRulesComparator(sortingOptions))
+        .slice((page - 1) * perPage, page * perPage)
+    : filteredRules;
+
+  const queryClient = useQueryClient();
+  /**
+   * Use this method to update rules data cached by react-query.
+   * It is useful when we receive new rules back from a mutation query (bulk edit, etc.);
+   * we can merge those rules with the existing cache to avoid an extra roundtrip to re-fetch updated rules.
+   */
+  const updateRules = useCallback(
+    (newRules: Rule[]) => {
+      queryClient.setQueryData<ReturnType<typeof useAllRules>['data']>(
+        getAllRulesQueryKey({ isInMemorySorting, filterOptions, sortingOptions, pagination }),
+        (currentData) => ({
+          rules: mergeRules(currentData?.rules || [], newRules),
+          total: currentData?.total || 0,
+        })
+      );
+
+      /**
+       * Unset loading state for all new rules
+       */
+      const newRuleIds = newRules.map((r) => r.id);
+      const newLoadingRuleIds = loadingRules.ids.filter((id) => !newRuleIds.includes(id));
+      setLoadingRules({
+        ids: newLoadingRuleIds,
+        action: newLoadingRuleIds.length === 0 ? null : loadingRules.action,
+      });
+    },
+    [
+      filterOptions,
+      isInMemorySorting,
+      loadingRules.action,
+      loadingRules.ids,
+      pagination,
+      queryClient,
+      sortingOptions,
+    ]
+  );
+
+  return (
+    <RulesTableContext.Provider
+      value={{
```

Comment:
Should we memoize `value` to prevent re-rendering of context consumers? As it passed currently as a new object in provider, every time context re-renders itself.

Reply:
Done